### PR TITLE
bpo-41972: Use the two-way algorithm for string searching

### PR DESCRIPTION
**Pull request by sweeneyde**
https://bugs.python.org/issue41972

### CODE REVIEW

File: Objects/stringlib/fastsearch.h [hunk 1]
```diff
@@ -160,6 +164,379 @@ STRINGLIB(rfind_char)(const STRINGLIB_CHAR* s, Py_ssize_t n, STRINGLIB_CHAR ch)
...
+    // Now fill up a table
+    memset(&(p->table[0]), 0xff, TABLE_SIZE*sizeof(SHIFT_TYPE));
+    assert(p->table[0] == NOT_FOUND);
+    assert(p->table[TABLE_MASK] == NOT_FOUND);
+    for (Py_ssize_t i = 0; i < len_needle; i++) {
+        Py_ssize_t shift = len_needle - i;
+        if (shift > SHIFT_OVERFLOW) {
+            shift = SHIFT_OVERFLOW;
+        }
+        p->table[needle[i] & TABLE_MASK] = shift;
```
Comment (tim-one) [inline Objects/stringlib/fastsearch.h]:
```suggestion
        p->table[needle[i] & TABLE_MASK] = Py_SAFE_DOWNCAST(shift,
                                                    Py_ssize_t, SHIFT_TYPE);
```

Implicit narrowing conversions generate compiler warnings under MSVC. The edit stops those, and, in debug mode, will assert-fail if bits are actually lost.

File: Objects/stringlib/fastsearch.h [hunk 2]
```diff
@@ -9,10 +9,14 @@
...
    Python's ordinary string types, but may cause problems if you're
    using this code in other contexts.  also, the count mode returns -1
-   if there cannot possible be a match in the target string, and 0 if
+   if there cannot possibly be a match in the target string, and 0 if
    it has actually checked for matches, but didn't find any.  callers
    beware! */
 
+/* If the needle is long enough, use Crochemore and Perrin's Two-Way
+   algorithm, which has guaranteed O(n) runtime. Also compute a table
+   of shifts to sometimes achieve O(n/m) runtime in the best cases. */
```
Comment (tim-one) [inline Objects/stringlib/fastsearch.h]: For the last two lines, suggest instead:

> algorithm, which has worst-case O(n) runtime, and best case O(n/k).  Also compute a table of shifts to achieve O(n/k) in more cases, and to often (data dependent) deduce larger shifts than pure C&P can deduce.

**Comment by sweeneyde** [comment]
@github-actions I think this should stay open

**Comment by tim-one** [comment]
Absolutely keep it open! I've been buried under other stuff, but am still keen to get this merged.

**Comment by Mariatta** [comment]
We've configured it so that GitHub action doesn't close the PR, so don't worry. I will see about adjusting that message.

**Comment by gvanrossum** [comment]
If Tim approves we might get this into alpha 6 which goes out Monday. @pablogsal what do you think?

**Review by tim-one — APPROVED** [review]
As briefly explained on BPO, I'd very much like to see this get in, but can't make the time needed for an in-depth review for the foreseeable future. So I'm approving this based on the last serious review I did (some months ago, alas). It would be great to get a more recent review, though.

**Comment by zanek2019** [comment]
How do you create apps in python?

**Comment by pablogsal** [comment]
> If @tim-one  approves we might get this into alpha 6 which goes out Monday. @pablogsal what do you think?

Yup, it should not be any problem. As I have a SC meeting on Monday I will probably start the release a bit later than usual (around 7PM London time).

**Comment by tim-one** [comment]
To be explicit here, am I "it" for hitting the "Squash and merge" button if/when the time comes for that? Fine by me if so, just want to be clear so it doesn't get dropped due to mistaken assumptions.

If I am "it", and don't hear more, I'll hit the button tomorrow afternoon (about 18 hours from now, approximately 19:00 CST Saturday, so about 13:00 CST Sunday).

**Comment by gvanrossum** [comment]
W00t! This deserves a note in what's new as well (Doc/whatsnew/3.10.rst).

**Pull request merged**